### PR TITLE
CryptoCpp facade

### PR DESCRIPTION
**Pull request by bartekryba**
CryptoCpp and starknet-rs facade

Uses: 
- starknet cryptocpp - ccryptocpp.xcframework generated from [crypto-cpp-swift](https://github.com/software-mansion-labs/crypto-cpp-swift) (Currently compiling only for arm64 ios and iossimulator, will add x86_64 simulator soon)
- `generate_k` c binding from [starknet-rs](https://github.com/bartekryba/starknet-rs-c-bindings) - rust module compiled into static library and made into xcframework.

### CODE REVIEW

File: Sources/CryptoToolkit/Utils.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+internal let bufferByteSize = 32
+
+public enum CryptoToolkitError: Error {
+    case cryptoCppError
+    case secp256K1Error
+}
+
+internal extension Data {
+    func paddedLeftUpTo(_ length: Int) -> Data {
```
Comment (Solpatium) [inline Sources/CryptoToolkit/Utils.swift:26]: `padLeft`?

File: Sources/CCryptopp/rfc6979.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    auto dsaKCalculator = DL_Algorithm_DSA_RFC6979<ECP, SHA256>();
+    
+    auto privateKeyInt = Integer(privateKey, 32);
+    auto subgroupGeneratorInt = Integer(subgroupGenerator, 32);
+    auto hashInt = Integer(hash, 32);
+    
+    Integer k;
+    
+    for (int i = 0; i <= attempt; i++) {
+        // Note that, contrary to it's name, it generates k in a deterministic way.
```
Comment (Solpatium) [inline Sources/CCryptopp/rfc6979.cpp:25]: its

File: Sources/Starknet/Crypto/StarknetCurve.swift [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        if let lastError = lastError {
+            throw lastError
+        }
+        
+        throw StarknetCurveError.unknownError
+    }
+}
+
+fileprivate extension Data {
+    func toFelt() -> Felt? {
```
Comment (Solpatium) [inline Sources/Starknet/Crypto/StarknetCurve.swift:91]: I guess this can throw right away since you use it this way above.

**Review by Solpatium — CHANGES_REQUESTED** [review]

File: Sources/Starknet/Crypto/StarknetCurve.swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    
+    public class func verify(publicKey: Felt, hash: Felt, r: Felt, s: Felt) throws -> Bool {
+        guard let w = s.value.inverse(curveOrder) else {
+            throw StarknetCurveError.verifyError
+        }
+        
+        return try CryptoCpp.verify(publicKey: publicKey.serialize(), hash: hash.serialize(), r: r.serialize(), s: w.serialize())
+    }
+    
+    internal class func sign(privateKey: Felt, hash: Felt, k: BigUInt) throws -> StarknetCurveSignature {
```
Comment (Solpatium) [inline Sources/Starknet/Crypto/StarknetCurve.swift:57]: Please add comment explaining why this is internal.

File: Sources/CryptoToolkit/Utils.swift (hunk 1 above)
Comment (bartekryba) [reply Sources/CryptoToolkit/Utils.swift]: Renamed it to paddingLeft, according to how apple names non-mutating methods

File: Sources/Starknet/Crypto/StarknetCurve.swift [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import BigInt
+
+public enum StarknetCurveError: Error {
+    case deserializationError
+    case verifyError
+    case unknownError
+}
+
+public class StarknetCurve {
+    private static let curveOrder = BigUInt("800000000000010FFFFFFFFFFFFFFFFB781126DCAE7B2321E66A241ADC64D2F", radix: 16)!
```
Comment (dmcrodrigues) [inline Sources/Starknet/Crypto/StarknetCurve.swift:12]: Can we make this public? In the jvm it is public and we need it to do the [account derivation](https://community.starknet.io/t/account-keys-and-addresses-derivation-standard/1230), more specifically to implement the grinding of the key.

File: Sources/Starknet/Crypto/StarknetCurve.swift [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    case deserializationError
+    case verifyError
+    case unknownError
+}
+
+public class StarknetCurve {
+    /// x coordinate of Starknet curve generator.
+    public static let curveOrder = BigUInt("800000000000010FFFFFFFFFFFFFFFFB781126DCAE7B2321E66A241ADC64D2F", radix: 16)!
+    
+    /// Compute pedersen hash on imput values.
```
Comment (Solpatium) [inline Sources/Starknet/Crypto/StarknetCurve.swift:15]: imput 
image

**Review by Solpatium — APPROVED** [review]

**Review by THenry14 — APPROVED** [review]

**Pull request merged**